ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/FollowerPromptness.hs

```diff
@@ -0,0 +1,214 @@
+                  ChainDBImpl.OutdatedTentativeHeader hdr ->
+                    addTiming varTentativeUnsetTimings (headerHash hdr)
+                  ChainDBImpl.TrapTentativeHeader hdr ->
+                    addTiming varTentativeUnsetTimings (headerHash hdr)
+                  _ -> pure ()
+                _ -> pure ()
+            _ -> pure ()
+    chainDB <- openChainDB registry chainDBTracer
+
+    -- Continually fetch instructions from a tentative follower.
+    follower <-
+      ChainDB.newFollower chainDB registry ChainDB.TentativeChain ChainDB.GetHash
+    _ <- forkLinkedThread registry "Follower listener" $ forever $
+      ChainDB.followerInstructionBlocking follower >>= \case
+       Chain.AddBlock hdrHash -> addTiming varFollowerInstrTimings hdrHash
+       Chain.RollBack _       -> pure ()
+
+    -- Add all blocks to the ChainDB.
+    let addBlock = ChainDB.addBlock_ chainDB Punishment.noPunishment
+    for_ chainUpdates \case
+      AddBlock blk      -> addBlock blk
+      SwitchFork _ blks -> for_ blks addBlock
+
+    tentativeHeaderSetTimings   <- readTVarIO varTentativeSetTimings
+    tentativeHeaderUnsetTimings <- readTVarIO varTentativeUnsetTimings
+    followerInstrTimings        <- readTVarIO varFollowerInstrTimings
+    traceByTime                 <- getTrace
+    pure FollowerPromptnessOutcome {..}
+  where
+    openChainDB ::
+         ResourceRegistry m
+      -> Tracer m (ChainDBImpl.TraceEvent TestBlock)
+      -> m (ChainDB m TestBlock)
+    openChainDB registry cdbTracer = do
+        chainDbArgs <- do
+          let mcdbTopLevelConfig = singleNodeTestConfigWithK securityParam
+              mcdbChunkInfo      = mkTestChunkInfo mcdbTopLevelConfig
+              mcdbInitLedger     = testInitExtLedger
+              mcdbRegistry       = registry
+          mcdbNodeDBs <- emptyNodeDBs
+          let cdbArgs = fromMinimalChainDbArgs MinimalChainDbArgs{..}
+          pure $ cdbArgs { cdbTracer = cdbTracer }
+        (_, (chainDB, ChainDBImpl.Internal{intAddBlockRunner})) <-
+          allocate
+            registry
+            (\_ -> ChainDBImpl.openDBInternal chainDbArgs False)
+            (ChainDB.closeDB . fst)
+        _ <- forkLinkedThread registry "AddBlockRunner" intAddBlockRunner
+        pure chainDB
+
+    withTime = contramapM \ev -> (,ev) <$> getMonotonicTime
+
+    addTiming varTiming hash = do
+        now <- getMonotonicTime
+        atomically $ modifyTVar varTiming $
+          Map.alter (Just . maybe (Set.singleton hash) (Set.insert hash)) now
+
+data FollowerPromptnessTestSetup = FollowerPromptnessTestSetup {
+    chainUpdates :: [ChainUpdate]
+  }
+  deriving stock (Show)
+
+instance Condense FollowerPromptnessTestSetup where
+  condense FollowerPromptnessTestSetup{..} =
+    "Chain updates: " <> condense chainUpdates
+
+instance Arbitrary FollowerPromptnessTestSetup where
+  arbitrary = do
+      chainUpdates <- genChainUpdates TentativeChainBehavior securityParam 20
```

Comment:
Are there any properties of `genChainUpdates` that we're assuming here?

I just noticed it doesn't have a comment on it :(

EG, at a glance, it looks like it can generate `SwitchFork n bs` where `length bs == n`, but I didn't see anything ensuring that in that case the `last bs` is preferred to the youngest rolled back block. Thus these `ChainUpdate`s do not necessarily monotonically improve the chain tip 🤔

That might not matter to this test, but perhaps a one line comment here clarifying the extent to which it does or doesn't would be worthwhile.

Reply:
Yes, it is not guaranteed that every chain update will result in a tentative header, but this is not important for this test, as it still happens sufficiently often. I am adding a comment stressing this :+1:

Reply:
Discussed on a call.

- Esgen pointed out that this `TestBlock(With)` uses the `Bft` protocol, which has no tie-breaker. And so `length bs == n` will never cause a switch (with no tiebreaker, the behavior is just to keep the chain you already have).

- We'll open a ticket to enrich other tentative header tests specifically with interesting tiebreaker behaviors. But for this promptness test, its a bit off-topic: `AddBlock` will trigger the tentative pathway.